frontend/src/pages/Compare.tsx

```diff
@@ -28,16 +28,27 @@ import { PageProps } from './Page';
 
 export type CompareProps = PageProps & CompareState;
 
+enum CompareVersion {
+  V1,
+  V2,
+  Mixed,
+  InvalidRunCount,
+}
+
 // This is a router to determine whether to show V1 or V2 compare page.
 export default function Compare(props: CompareProps) {
+  const [compareVersion, setCompareVersion] = useState<CompareVersion>(
+    CompareVersion.InvalidRunCount,
+  );
   const queryParamRunIds = new URLParser(props).get(QUERY_PARAMS.runlist);
   const runIds = (queryParamRunIds && queryParamRunIds.split(',')) || [];
 
-  // Retrieves run details.
-  const { isSuccess, isError, data } = useQuery<ApiRunDetail[], Error>(
+  // Retrieves run details, set page version on success.
+  const { isError } = useQuery<ApiRunDetail[], Error>(
     ['run_details', { ids: runIds }],
     () => Promise.all(runIds.map(async id => await Apis.runServiceApi.getRun(id))),
     {
+      refetchOnMount: 'always',
```

Comment:
I actually realized that if there is any logic in an `onSuccess` callback, then this is necessary; otherwise, if the query is old, that `onSuccess` callback will never be called.

In my case, I initially set the `compareVersion` within that callback; however, I just updated the code so that `useQuery` returns `data` and I set the `compareVersion` in a separate `useEffect` hook: https://github.com/kubeflow/pipelines/pull/7844/commits/1a341f9d5b543851e6e0b03e10572175713f5339.

This is helpful, as it both speeds up the page (`data` that has already been fetched) and prevents unnecessary requests. The only potential downside is that if new information is available about run versions which is pertinent to the `Compare.tsx` page, it will require a manual user refresh. I think that it is unlikely that run versions would change from V1 to V2 during a user session, so I think this design is OK, but let me know if you have other thoughts.

Reply:
Thanks for the investigation! I agree with using `useEffect` to replace `onSuccess` to avoid unnecessary refetch when possible. refetching new information in runVersion is helpful, but it should be triggered via user action (Hit refresh button) to maintain the stability of Compare page.